File: apps/web/lib/paragraphs.ts
/** 阅读段落边界：CNV 生成表 + 离线兜底算法（对齐 scripts/build_paragraphs.py）。 */

export interface VerseLine {
  verse: number;
  text: string;
}

export interface VerseParagraph {
  startVerse: number;
  endVerse: number;
  verses: VerseLine[];
}

export type ParagraphRange = [number, number];

const POETRY_BOOKS = new Set([
  'PSA', 'PRO', 'ECC', 'SNG', 'LAM', 'AMO', 'MIC', 'HAB', 'ZEP', 'NAH',
  'HAG', 'ZEC', 'MAL', 'JOB',
]);

const MIN_VERSES = 2;
const MAX_VERSES = 6;
const MAX_CHARS = 320;
const MIN_WEAK_VERSES = 3;
const MIN_WEAK_CHARS = 120;

export function isPoetryBook(bookId: string): boolean {
  return POETRY_BOOKS.has(bookId.toUpperCase());
}

function endsSentence(text: string): boolean {
  return /[。！？；….!?;:]["'」』)]*$/.test(text.trim());
}

function charCount(buf: VerseLine[]): number {
  return buf.reduce((n, v) => n + v.text.length, 0);
}

function mergeSingletonRanges(
  ranges: ParagraphRange[],
  verseMap: Map<number, VerseLine>,
): ParagraphRange[] {
  if (ranges.length <= 1) return ranges;
  const out: ParagraphRange[] = [];
  for (const [start, end] of ranges) {
    if (end - start + 1 > 1 || !out.length) {
      out.push([start, end]);
      continue;
    }
    const prev = out[out.length - 1]!;
    const combined = prev[1] - prev[0] + 2;
    let chars = 0;
    for (let n = prev[0]; n <= end; n++) {
      chars += verseMap.get(n)?.text.length ?? 0;
    }
    if (combined <= MAX_VERSES && chars <= MAX_CHARS) {
      prev[1] = end;
    } else {
      out.push([start, end]);
    }
  }
  return out;
}

function segmentStarts(sectionStarts: number[], first: number, last: number): number[] {
  const starts = [...new Set(sectionStarts.filter((s) => s >= first && s <= last))].sort(
    (a, b) => a - b,
  );
  if (!starts.length || starts[0] !== first) {
    return [first, ...starts.filter((s) => s !== first)];
  }
  return starts;
}

function groupSegment(verses: VerseLine[]): ParagraphRange[] {
  if (!verses.length) return [];
  const ranges: ParagraphRange[] = [];
  let buf: VerseLine[] = [];

  const flush = () => {
    if (!buf.length) return;
    ranges.push([buf[0]!.verse, buf[buf.length - 1]!.verse]);
    buf = [];
  };

  for (const v of verses) {
    if (buf.length) {
      if (buf.length >= MAX_VERSES || charCount(buf) >= MAX_CHARS) {
        flush();
      } else if (
        buf.length >= MIN_WEAK_VERSES
        && charCount(buf) >= MIN_WEAK_CHARS
        && endsSentence(buf[buf.length - 1]!.text)
      ) {
        flush();
      }
    }
    buf.push(v);
  }
  flush();

  const verseMap = new Map(verses.map((v) => [v.verse, v]));
  return mergeSingletonRanges(ranges, verseMap);
}

/** 离线兜底：sections 小标题 + 合并规则（无 paragraphs.json 时）。 */
export function computeParagraphRanges(
  bookId: string,
  verses: VerseLine[],
  sectionStarts: number[] = [],
): ParagraphRange[] {
  if (!verses.length) return [];
  if (isPoetryBook(bookId)) {
    return verses.map((v) => [v.verse, v.verse] as ParagraphRange);
  }

  const sorted = [...verses].sort((a, b) => a.verse - b.verse);
  const first = sorted[0]!.verse;
  const last = sorted[sorted.length - 1]!.verse;
  const starts = segmentStarts(sectionStarts, first, last);
  const out: ParagraphRange[] = [];

  for (let i = 0; i < starts.length; i++) {
    const segStart = starts[i]!;
    const segEnd = i + 1 < starts.length ? starts[i + 1]! - 1 : last;
    const segVerses = sorted.filter((v) => v.verse >= segStart && v.verse <= segEnd);
    out.push(...groupSegment(segVerses));
  }
  return out;
}

export function paragraphsFromRanges(
  verses: VerseLine[],
  ranges: ParagraphRange[],
): VerseParagraph[] {
  if (!verses.length || !ranges.length) return [];
  const map = new Map(verses.map((v) => [v.verse, v]));
  const out: VerseParagraph[] = [];
  for (const [start, end] of ranges) {
    const chunk: VerseLine[] = [];
    for (let n = start; n <= end; n++) {
      const v = map.get(n);
      if (v) chunk.push(v);
    }
    if (chunk.length) {
      out.push({ startVerse: chunk[0]!.verse, endVerse: chunk[chunk.length - 1]!.verse, verses: chunk });
    }
  }
  return out;
}

/** 将一章经节分组为连续段落。优先使用出版段落表，否则走兜底算法。 */
export function groupVersesIntoParagraphs(
  bookId: string,
  verses: VerseLine[],
  sectionStarts: number[] = [],
  paragraphRanges?: ParagraphRange[] | null,
): VerseParagraph[] {
  if (!verses.length) return [];
  const ranges =
    paragraphRanges && paragraphRanges.length
      ? paragraphRanges
      : computeParagraphRanges(bookId, verses, sectionStarts);
  return paragraphsFromRanges(verses, ranges);
}

File: apps/mobile/lib/features/bible/paragraphs.dart
/// 阅读段落边界：CNV 段落表 + 离线兜底算法（对齐 PWA `paragraphs.ts`）。
library;

import 'models.dart';

class VerseParagraph {
  VerseParagraph({
    required this.startVerse,
    required this.endVerse,
    required this.verses,
  });

  final int startVerse;
  final int endVerse;
  final List<Verse> verses;
}

typedef ParagraphRange = (int start, int end);

/// 散文段首缩进（RichText 无 text-indent，1 个全角字宽）。
const kProseParagraphIndent = '\u3000';

const _poetryBooks = {
  'PSA', 'PRO', 'ECC', 'SNG', 'LAM', 'AMO', 'MIC', 'HAB', 'ZEP', 'NAH',
  'HAG', 'ZEC', 'MAL', 'JOB',
};

const _minVerses = 2;
const _maxVerses = 6;
const _maxChars = 320;
const _minWeakVerses = 3;
const _minWeakChars = 120;

bool isPoetryBook(String bookId) =>
    _poetryBooks.contains(bookId.toUpperCase());

bool _endsSentence(String text) {
  final t = text.trim();
  if (t.isEmpty) return false;
  return RegExp('[。！？；….!?;:]["\'\u300d\u300f)]*\$').hasMatch(t);
}

int _charCount(List<Verse> buf) =>
    buf.fold<int>(0, (n, v) => n + v.text.length);

List<ParagraphRange> _mergeSingletonRanges(
  List<ParagraphRange> ranges,
  Map<int, Verse> verseMap,
) {
  if (ranges.length <= 1) return ranges;
  final out = <ParagraphRange>[];
  for (final (start, end) in ranges) {
    if (end - start + 1 > 1 || out.isEmpty) {
      out.add((start, end));
      continue;
    }
    final prev = out.last;
    final combined = prev.$2 - prev.$1 + 2;
    var chars = 0;
    for (var n = prev.$1; n <= end; n++) {
      chars += verseMap[n]?.text.length ?? 0;
    }
    if (combined <= _maxVerses && chars <= _maxChars) {
      out[out.length - 1] = (prev.$1, end);
    } else {
      out.add((start, end));
    }
  }
  return out;
}

List<int> _segmentStarts(List<int> sectionStarts, int first, int last) {
  final starts = sectionStarts.where((s) => s >= first && s <= last).toSet().toList()
    ..sort();
  if (starts.isEmpty || starts.first != first) {
    return [first, ...starts.where((s) => s != first)];
  }
  return starts;
}

List<ParagraphRange> _groupSegment(List<Verse> verses) {
  if (verses.isEmpty) return [];
  final ranges = <ParagraphRange>[];
  var buf = <Verse>[];

  void flush() {
    if (buf.isEmpty) return;
    ranges.add((buf.first.verse, buf.last.verse));
    buf = [];
  }

  for (final v in verses) {
    if (buf.isNotEmpty) {
      if (buf.length >= _maxVerses || _charCount(buf) >= _maxChars) {
        flush();
      } else if (buf.length >= _minWeakVerses &&
          _charCount(buf) >= _minWeakChars &&
          _endsSentence(buf.last.text)) {
        flush();
      }
    }
    buf.add(v);
  }
  flush();

  final verseMap = {for (final v in verses) v.verse: v};
  return _mergeSingletonRanges(ranges, verseMap);
}

/// 离线兜底：sections 小标题 + 合并规则。
List<ParagraphRange> computeParagraphRanges(
  String bookId,
  List<Verse> verses, [
  List<int> sectionStarts = const [],
]) {
  if (verses.isEmpty) return [];
  if (isPoetryBook(bookId)) {
    return verses.map((v) => (v.verse, v.verse)).toList();
  }

  final sorted = [...verses]..sort((a, b) => a.verse.compareTo(b.verse));
  final first = sorted.first.verse;
  final last = sorted.last.verse;
  final starts = _segmentStarts(sectionStarts, first, last);
  final out = <ParagraphRange>[];

  for (var i = 0; i < starts.length; i++) {
    final segStart = starts[i];
    final segEnd = i + 1 < starts.length ? starts[i + 1] - 1 : last;
    final segVerses =
        sorted.where((v) => v.verse >= segStart && v.verse <= segEnd).toList();
    out.addAll(_groupSegment(segVerses));
  }
  return out;
}

List<VerseParagraph> paragraphsFromRanges(
  List<Verse> verses,
  List<ParagraphRange> ranges,
) {
  if (verses.isEmpty || ranges.isEmpty) return [];
  final map = {for (final v in verses) v.verse: v};
  final out = <VerseParagraph>[];
  for (final (start, end) in ranges) {
    final chunk = <Verse>[];
    for (var n = start; n <= end; n++) {
      final v = map[n];
      if (v != null) chunk.add(v);
    }
    if (chunk.isNotEmpty) {
      out.add(VerseParagraph(
        startVerse: chunk.first.verse,
        endVerse: chunk.last.verse,
        verses: chunk,
      ));
    }
  }
  return out;
}

List<VerseParagraph> groupVersesIntoParagraphs(
  String bookId,
  List<Verse> verses, {
  List<int> sectionStarts = const [],
  List<ParagraphRange>? paragraphRanges,
}) {
  if (verses.isEmpty) return [];
  final ranges = (paragraphRanges != null && paragraphRanges.isNotEmpty)
      ? paragraphRanges
      : computeParagraphRanges(bookId, verses, sectionStarts);
  return paragraphsFromRanges(verses, ranges);
}

List<ParagraphRange> parseParagraphRangesJson(List<dynamic> raw) {
  return raw
      .map((e) {
        if (e is! List || e.length < 2) return null;
        final a = (e[0] as num?)?.toInt();
        final b = (e[1] as num?)?.toInt();
        if (a == null || b == null) return null;
        return (a, b);
      })
      .whereType<ParagraphRange>()
      .toList();
}
